sdk/eidos_sdk/cpu/conversational_agent_cpu.py

```diff
@@ -59,56 +58,113 @@ async def schedule_request(
             call_context: CallContext,
             prompts: List[CPUMessageTypes],
             output_format: Union[Literal["str"], Dict[str, Any]] = "str",
-    ) -> Any:
+    ) -> AsyncGenerator[StreamEvent, None]:
         try:
             conversation_messages = await self.io_unit.process_request(prompts)
-            conversation = await self.memory_unit.storeAndFetch(call_context, conversation_messages)
-            assistant_message = await self._llm_execution_cycle(call_context, conversation, output_format)
-            return await self.io_unit.process_response(call_context, assistant_message.content)
+            await self.memory_unit.storeMessages(call_context, conversation_messages)
+            context = f"{call_context.process_id}"
+            if call_context.thread_id:
+                context += f":{call_context.thread_id}"
+            llm_it = with_context(context, self._llm_execution_cycle(call_context, output_format))
+            async for event in llm_it:
+                yield event
         except HTTPException:
             raise
         except Exception as e:
-            raise RuntimeError("Error in cpu while processing request") from e
+            yield ErrorEvent(reason=str(e))
+
+    async def _call_tool(self, call_context: CallContext, tool_call_event: ToolCallEvent, tool_defs):
+        parent_stream_context = tool_call_event.stream_context
+        tool_def = tool_defs[tool_call_event.tool_name]
+        tool_context = [*tool_call_event.stream_context, tool_call_event.tool_call_id]
+        try:
+            tc = ToolCall(
+                tool_call_id=tool_call_event.tool_call_id,
+                name=tool_call_event.tool_name,
+                arguments=tool_call_event.tool_args,
+            )
+            yield ToolCallEvent(stream_context=tool_context,
+                                tool_call_id=tool_call_event.tool_call_id,
+                                tool_name=tool_call_event.tool_name,
+                                tool_args=tool_call_event.tool_args)
+
+            # todo -- change this into a stream on output
```

Comment:
Do you think we should support both streaming and non-streaming tools? Ie normal returns for simple tools?